man/tss2_sign.1.md
% tss2_sign(1) tpm2-tools | General Commands Manual
%
% APRIL 2019

# NAME

**tss2_sign**(1) -

# SYNOPSIS

**tss2_sign** [*OPTIONS*]

[common fapi references](common/tss2-fapi-references.md)

# DESCRIPTION

**tss2_sign**(1) - This command uses a key inside the TPM to sign a digest value
using the TPM signing schemes as specified in the cryptographic profile
(cf., **fapi-profile(5)**). The digest can be provided as an argument
or computed from data passed as an argument.

# OPTIONS

These are the available options:

  * **-p**, **\--keyPath**=_STRING_:

    The path to the signing key.

  * **-s**, **\--padding**=_STRING_:

    The padding scheme used. Possible values are "RSA_SSA", "RSA_PSS" (case insensitive). Optional parameter.
    If omitted, the default padding specified in the cryptographic profile
    (cf., **fapi-profile(5)**) is used.

  * **-c**, **\--certificate**=_FILENAME_ or _-_ (for stdout):

    The certificate associated with keyPath in PEM format. Optional parameter.

  * **-d**, **\--digest**=_FILENAME_ or _-_ (for stdin):

    The data to be signed, already hashed.

  * **-m**, **\--data**=_FILENAME_ or _-_ (for stdin):

    The data from which the hash is computed, which is then signed.

  * **-f**, **\--force**:

    Force overwriting the output file.

  * **-k**, **\--publicKey**=_FILENAME_ or _-_ (for stdout):

    The public key associated with keyPath in PEM format. Optional parameter.

  * **-o**, **\--signature**=_FILENAME_ or _-_ (for stdout):

    Returns the signature in binary form.

[common tss2 options](common/tss2-options.md)

# EXAMPLE

```
tss2_sign --keyPath=HS/SRK/myRSASign --padding="RSA_PSS" --digest=digest.file --signature=signature.file --publicKey=publicKey.file
```

# RETURNS

0 on success or 1 on failure.

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/fapi/tss2_sign.c
/* SPDX-License-Identifier: BSD-3-Clause */

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "tools/fapi/tss2_template.h"

/* Context struct used to store passed commandline parameters */
static struct cxt {
    char const *keyPath;
    char const *digest;
    char const *data;
    char const *signature;
    char const *publicKey;
    char const *certificate;
    bool        overwrite;
    char const *padding;
} ctx;

/* Parse command line parameters */
static bool on_option(char key, char *value) {
    switch (key) {
    case 'c':
        ctx.certificate = value;
        break;
    case 'd':
        ctx.digest = value;
        break;
    case 'm':
        ctx.data = value;
        break;
    case 'f':
        ctx.overwrite = true;
        break;
    case 'p':
        ctx.keyPath = value;
        break;
    case 'k':
        ctx.publicKey = value;
        break;
    case 'o':
        ctx.signature = value;
        break;
    case 's':
        ctx.padding = value;
        break;
    }
    return true;
}

/* Define possible command line parameters */
static bool tss2_tool_onstart(tpm2_options **opts) {
    struct option topts[] = {
        {"keyPath",     required_argument, NULL, 'p'},
        {"padding",     required_argument, NULL, 's'},
        {"digest",      required_argument, NULL, 'd'},
        {"data",        required_argument, NULL, 'm'},
        {"signature",   required_argument, NULL, 'o'},
        {"publicKey",   required_argument, NULL, 'k'},
        {"force",       no_argument      , NULL, 'f'},
        {"certificate", required_argument, NULL, 'c'},

    };
    return (*opts = tpm2_options_new ("c:d:m:fp:k:o:s:", ARRAY_LEN(topts), topts,
                                      on_option, NULL, 0)) != NULL;
}

/* Execute specific tool */
static int tss2_tool_onrun (FAPI_CONTEXT *fctx) {

    /* Check availability of required parameters */

#ifndef HAVE_FAPI_DIGEST_AND_SIGN
    if (ctx.data) {
        fprintf (stderr, "Fapi_DigestAndSign not available in the current FAPI version.\n");
        return -1;
    }
#endif
    if (!ctx.digest && !ctx.data) {
        fprintf (stderr, "digest or dataa missing, use --digest or --data\n");
        return -1;
    }
    if (ctx.digest && ctx.data) {
        fprintf (stderr, "use --digest or --data\n");
        return -1;
    }
    if (!ctx.keyPath) {
        fprintf (stderr, "key path missing, use --keyPath\n");
        return -1;
    }
    if (!ctx.signature) {
        fprintf (stderr, "signature missing, use --signature\n");
        return -1;
    }

    /* Check exclusive access to stdout */
    int count_out = 0;
    if (ctx.certificate && !strcmp (ctx.certificate, "-")) count_out +=1;
    if (ctx.signature && !strcmp (ctx.signature, "-")) count_out +=1;
    if (ctx.publicKey && !strcmp (ctx.publicKey, "-")) count_out +=1;
    if (count_out > 1) {
        fprintf (stderr, "Only one of --certificate, --signature and "\
        "--publicKey can print to - (standard output)\n");
        return -1;
    }

    /* Read data needed to create signature */
    uint8_t *data = NULL, *digest = NULL, *signature = NULL;
    size_t dataSize = 0, digestSize = 0, signatureSize;
    char *publicKey = NULL, *certificate = NULL;
    TSS2_RC r;
    if (ctx.digest) {
        r = open_read_and_close (ctx.digest, (void**)&digest, &digestSize);
    } else {
        r = open_read_and_close (ctx.data, (void**)&data, &dataSize);
    }

    if (r){
        return 1;
    }

    /* Execute FAPI command with passed arguments */
    if (ctx.digest) {
        r = Fapi_Sign (fctx, ctx.keyPath, ctx.padding, digest,
        digestSize, &signature, &signatureSize, &publicKey, &certificate);
        if (r != TSS2_RC_SUCCESS) {
            LOG_PERR ("Fapi_Sign", r);
            free (digest);
            return 1;
        }
        free (digest);
    }
#ifdef HAVE_FAPI_DIGEST_AND_SIGN
    else if (ctx.data) {
        r = Fapi_DigestAndSign (fctx, ctx.keyPath, ctx.padding, data,
        dataSize, &signature, &signatureSize, &publicKey, &certificate);
        if (r != TSS2_RC_SUCCESS) {
            LOG_PERR ("Fapi_Sign", r);
            free (data);
            return 1;
        }
        free (data);
    }
#endif

    /* Write returned data to file(s) */
    if (ctx.certificate && certificate && strlen(certificate)) {
            r = open_write_and_close (ctx.certificate, ctx.overwrite,
                certificate, strlen(certificate));
            if (r) {
                Fapi_Free (certificate);
                Fapi_Free (signature);
                Fapi_Free (publicKey);
                return 1;
            }
    }
    Fapi_Free (certificate);

    if (ctx.signature && signature) {
        r = open_write_and_close (ctx.signature, ctx.overwrite, signature,
            signatureSize);
        if (r) {
            Fapi_Free (signature);
            Fapi_Free (publicKey);
            return 1;
        }
    }
    Fapi_Free (signature);

    if (ctx.publicKey && publicKey) {
        r = open_write_and_close (ctx.publicKey, ctx.overwrite, publicKey,
                strlen(publicKey));
        if (r) {
            Fapi_Free (publicKey);
            return 1;
        }
    }
    Fapi_Free (publicKey);

    return 0;
}

TSS2_TOOL_REGISTER("sign", tss2_tool_onstart, tss2_tool_onrun, NULL)
